_drafts/2020-09-17-fixing-an-n1-performance-problem-in-xaf-xpo-with-totally-undocumented-apis.md

```diff
@@ -0,0 +1,948 @@
+19.09.20 11:55:45.372 Result: rowcount = 1000, total = 44000, N0.{Oid,Guid} = 16000, N0.{Hours,Int32} = 4000, N0.{SlowOfferWithPersistentAlias,Guid} = 16000, N0.{OptimisticLockField,Int32} = 4000, N0.{GCRecord,Int32} = 4000
+.
+.
+.
+19.09.20 11:55:46.007 Executing sql 'select N0."Oid",N0."Hours",N0."SlowOfferWithPersistentAlias",N0."OptimisticLockField",N0."GCRecord" from "dbo"."SlowOfferItemWithPersistentAlias" N0 where (N0."GCRecord" is null and (N0."SlowOfferWithPersistentAlias" = @p0))' with parameters {1a57b4dc-7812-48cc-83ae-15e184078c5f}
+19.09.20 11:55:46.019 Result: rowcount = 1000, total = 44000, N0.{Oid,Guid} = 16000, N0.{Hours,Int32} = 4000, N0.{SlowOfferWithPersistentAlias,Guid} = 16000, N0.{OptimisticLockField,Int32} = 4000, N0.{GCRecord,Int32} = 4000
+19.09.20 11:55:46.026 Executing sql 'select top 1 count(*) from "dbo"."SlowOfferWithPersistentAlias" N0 where N0."GCRecord" is null'
+19.09.20 11:55:46.027 Result: rowcount = 1, total = 4, SubQuery(Count,,) = 4
+19.09.20 11:55:46.028 Executing sql 'select top 1 count(*) from "dbo"."SlowOfferItemWithPersistentAlias" N0 where N0."GCRecord" is null'
+```
+
+This results in the following performance
+
+1. Sort by `Name`: 0.692 seconds
+2. Sort by `HourSum`: 0.741 seconds
+
+#### DataView
+
+In `DataView` mode only visible columns are fetched from the database, as well as including the sorting query and some support queries. All records will be fetched in one query
+
+```txt
+19.09.20 12:03:31.738 Executing sql 'select top 2147483647 (select sum(N1."Hours") as Res0 from "dbo"."SlowOfferItemWithPersistentAlias" N1 where ((N0."Oid" = N1."SlowOfferWithPersistentAlias") and N1."GCRecord" is null)),N0."Name",N0."Oid" from "dbo"."SlowOfferWithPersistentAlias" N0 where N0."GCRecord" is null'
+19.09.20 12:03:31.821 Result: rowcount = 300, total = 10848, SubQuery(Sum,N1.{Hours,Int32},Select N1.{Hours,Int32}
+  from "SlowOfferItemWithPersistentAlias" N1
+ where N0.{Oid,Guid} = N1.{SlowOfferWithPersistentAlias,Guid} And N1.{GCRecord,Int32} Is Null) = 1200, N0.{Name,String} = 4848, N0.{Oid,Guid} = 4800
+19.09.20 12:03:31.824 Executing sql 'select top 1 count(*) from "dbo"."SlowOfferWithPersistentAlias" N0 where N0."GCRecord" is null'
+19.09.20 12:03:31.825 Result: rowcount = 1, total = 4, SubQuery(Count,,) = 4
+19.09.20 12:03:31.825 Executing sql 'select top 1 count(*) from "dbo"."SlowOfferItemWithPersistentAlias" N0 where N0."GCRecord" is null'
+19.09.20 12:03:31.839 Result: rowcount = 1, total = 4, SubQuery(Count,,) = 4
+```
+
+This results in the following performance
+
+1. Sort by `Name`: 0.071 seconds
+2. Sort by `HourSum`: 0.078 seconds
+
+#### InstantFeedback
+
+In `InstantFeedback` mode records are loaded in an async manor. It operates almost the same as the `ServerMode` in terms of number of queries, but the user experience is much better, cause visible records will be fetched in the background.
```

Comment:
```suggestion
In `InstantFeedback` mode records are loaded in an asynchronously . It operates almost the same as the `ServerMode` in terms of number of queries, but the user experience is much better, because visible records will be fetched in the background.
```